### DISCUSSION TITLE
JEI compatibility

### DISCUSSION THREAD

**Discussion by Rearth** [Rearth/Oritech#14] [Todos]
Oritech currently only seems to work with JEI. Based on my JEI understanding, a similar implementation will need to be made for JEI. Or maybe there is a lib mod that allows recipes to be displayed in both?

**Comment by unilock** [comment]
If nobody else has offered already, I would be willing to implement [EMI](https://github.com/emilyploszaj/emi) support.  
(EMI can also load JEI plugins if both mods are installed, but I find EMI's API much easier to work with)

There's also this API for implementing a single plugin that works on both EMI and REI, but I've never used it, and have no idea how comprehensive it is: https://modrinth.com/mod/tla-api

**Reply by Rearth** [reply]
Hi, yes that'd be great, thank you. I think directly implementing EMI support would be the best idea here, not sure how future-proof the tla-api is at this point. Feel free to give it a try, noone else is working on it yet.

You can find the current REI compat implementation here: https://github.com/Rearth/Oritech/blob/master/src/main/java/rearth/oritech/init/compat/OritechREIPlugin.java

I think this could serve as a starting point to get into the relevant code sections. Although be aware that I never really dug deeper into the REI api and am probably doing a few things in less than ideal ways.

If you have any more questions, make sure to reach out here or via discord.

**Reply by unilock** [reply]
Alright, I have a work-in-progress branch adding EMI support here: https://github.com/unilock/Oritech/tree/emi

It's not quite finished yet - I'd like to make the EMI plugin class more similar to the REI plugin class in the way it registers its categories and workstations and such, for consistency, and currently fluid inputs / outputs aren't rendered at all. Actually, I haven't even started the game with it yet... but it should still be a good start.

owo-lib doesn't have an "adapter" for EMI like it does for REI, so that's making the port a bit more difficult than I expected :v

**Reply by Rearth** [reply]
That branch already starts to look like a great improvement. Feel free to open a PR once you think it's ready to be used

**Reply by unilock** [reply]
I've tidied up the plugin class and halfway implemented fluid rendering, and it seems to mostly work!

Here's an example of a more complex recipe UI:

<img src="https://github.com/Rearth/Oritech/assets/11802133/e7d59312-a16b-4f77-8e08-733a775a8cd3" width="50%">

As you can see, the "duration" text is misaligned, and fluids don't actually render where they should. That'll require more thinking. In the meantime, I'll open a draft PR.

**Reply by Rearth** [reply]
Hi, thanks again for the PR. I hope you don't mind that I took this topic over and finished the EMI integration. It's available on the main branch now. Since I'm not a too experienced github user I didn't create a branch of your branch and merge it into main but instead just directly merge it into my main branch.
I added a point credited to your work in the readme.
